### PR TITLE
Don't define builtins on newer clang versions in palrt2.h

### PR DESCRIPTION
**Pull request by hughbe**
Avoids errors such as `error: definition of builtin function '_InterlockedIncrement`

### CODE REVIEW

**Comment by delmyers** [comment]
Hi, @hugbe, it looks like you are trying to get Mac support for ClrInstrumentationEngine. If that's the case, we should probably have a larger discussion about that as a community since we don't have any validation for mac scenarios right now. The ClrInstrumentationEngine depends on an abstraction layer called the PAL, which we don't link to for Mac machines currently.

Also, it seems that this change might not be the correct one. @jander-msft @wiktork you might have more context, but it seems like we would probably need to do a little more refactoring to get this to work? I'm not sure.

**Comment by hughbe** [comment]
I made some cmake modifications to get it building on mac as a test.

However, I created a WSL Ubuntu 18.04 with latest version of clang and got these errors too. So this isn’t necessarily a fix for Mac

**Comment by hughbe** [comment]
> Also, it seems that this change might not be the correct one. @jander-msft @wiktork you might have more context, but it seems like we would probably need to do a little more refactoring to get this to work? I'm not sure.
 
I’ll try to detail what I’ve done here. Basically, at some point between clang 3.5 (which is the one we use to build this repo) and the latest version, clang added support for _Interlocked* intrinsics on Windows, mac, Linux etc.

Basically clang throws an error if you try to redefine a builtin. By guarding for this using #if !__has_builtin we keep the existing behaviour on older compilers and fix the build on newer ones. We can delete the code for _Interlocked* if we start supporting clang with minimum version that already has the built ins

**Comment by WilliamXieMSFT** [comment]
Spoke with @jander-msft offline, Clang 3.5 was used by the dotnet core libraries at the time we adopted their PAL implementation, so likely not able to upgrade any time soon. I'm fine with this change as a stop-gap, @hughbe, but you may encounter other issues down the line.

File: src/unix/inc/palrt2.h [hunk 1]
```diff
@@ -82,7 +92,7 @@ __inline int InlineIsEqualGUID(REFGUID rguid1, REFGUID rguid2)
 
 // --- winerror ---
 #define ERROR_INVALID_THREAD_ID          1444L
-#define ERROR_ELEVATION_REQUIRED         740L
+#define ERROR_ELEVATIONO_REQUIRED         740L
```
Comment (delmyers) [inline src/unix/inc/palrt2.h:95]: Looks like a typo?

**Comment by hughbe** [comment]
> Spoke with @jander-msft offline, Clang 3.5 was used by the dotnet core libraries at the time we adopted their PAL implementation, so likely not able to upgrade any time soon

I will note that Ubuntu 18.04 does not have clang 3.5 available (`sudo apt-get clang-3.5` Does not work)

File: src/unix/inc/palrt2.h (hunk 1 above)
Comment (hughbe) [reply src/unix/inc/palrt2.h:95]: Yup not sure how this snuck in

**Comment by hughbe** [comment]
Could you try triggering the build again? I suspect there may have been a sign issue where we casted ULONG to LONG. Not sure though. If that doesn't work, will see about `volatile`

**Comment by hughbe** [comment]
Looks like this worked :)

**Comment by WilliamXieMSFT** [comment]
@hughbe Looks like the builds fail with:

d:\a\1\s\src\InstrumentationEngine.Lib\refcount.cpp(245,69): error C2220: the following warning is treated as an error [d:\a\1\s\src\InstrumentationEngine.Lib\InstrumentationEngine.Lib.vcxproj]
##[warning]src\InstrumentationEngine.Lib\refcount.cpp(245,69): Warning C4197: 'volatile PLONG': top-level volatile in cast is ignored

**Comment by hughbe** [comment]
Fingers crossed... sorry about this

**Comment by delmyers** [comment]
> Fingers crossed... sorry about this

Trying the build again

**Pull request closed without merging**